src/dtls/network.c

```diff
@@ -0,0 +1,962 @@
+        pw->status = DPS_ERR_NETWORK;
+    }
+}
+
+static void TLSRead(DPS_NetConnection* cn)
+{
+    DPS_NetContext* netCtx = cn->node->netCtx;
+    int ret;
+
+    assert(IsNodeThread(cn->node));
+
+    ret = mbedtls_ssl_read(&cn->ssl, (unsigned char*)netCtx->plainBuffer, sizeof(netCtx->plainBuffer)-1);
+    if (ret == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+        DPS_DBGPRINT("TLSRead() connection was closed gracefully\n");
+        ret = 0;
+        // TODO: Mark the connection was closed.
+        // TODO: Callback with failure?
+    } else if (ret < 0) {
+        DPS_DBGPRINT("TLSRead() failed, mbedtls_ssl_read returned -0x%x\n\n", -ret);
+        // TODO: Mark the connection was closed.
+        // TODO: Callback with failure?
+    } else {
+        DPS_DBGPRINT("TLSRead() decrypted into %d bytes of plaintext\n", ret);
+
+#ifndef NDEBUG
+        for (int i = 0; i < ret; i++) {
+            if ((i % 16) == 0) printf("\n"); else printf("  ");
+            printf("%02X", (uint8_t)netCtx->plainBuffer[i]);
+        }
+        printf("\n\n");
+#endif
+
+        netCtx->receiveCB(netCtx->node, cn->peer, DPS_OK, (uint8_t*)netCtx->plainBuffer, ret);
+    }
+
+    memset(netCtx->plainBuffer, 0, sizeof(netCtx->plainBuffer));
+}
+
+static void TLSHandshake(DPS_NetConnection* cn)
+{
+    DPS_DBGTRACE();
+    assert(IsNodeThread(cn->node));
+
+    int ret = mbedtls_ssl_handshake(&cn->ssl);
+
+    if (ret == MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED) {
+        DPS_DBGPRINT("TLSHandshake() hello verification required, resetting\n");
+        mbedtls_ssl_session_reset(&cn->ssl);
+
+        if (cn->type == MBEDTLS_SSL_IS_SERVER) {
+            char clientID[128] = {};
+            const struct sockaddr* addr = (const struct sockaddr*)&cn->peer->addr.inaddr;
+            if (addr->sa_family == AF_INET) {
+                uv_ip4_name((const struct sockaddr_in*)addr, clientID, sizeof(clientID)-1);
+            } else {
+                uv_ip6_name((const struct sockaddr_in6*)addr, clientID, sizeof(clientID)-1);
+            }
+            ret = mbedtls_ssl_set_client_transport_id(&cn->ssl, (const unsigned char*)clientID, strlen(clientID));
+            if (ret != 0) {
+                DPS_ERRPRINT("ERROR: couldn't set client transport id (%d)\n", ret);
+            }
+        }
+
+        return;
+    }
+
+    // The two cases below just let us know that handshake is waiting for more
+    // data to be sent or received.
+    if (ret == MBEDTLS_ERR_SSL_WANT_READ) {
+        DPS_DBGPRINT("TLSHandshake() want read\n");
+        return;
+    }
+    if (ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+        DPS_DBGPRINT("TLSHandshake() want write\n");
+        return;
+    }
+
+    if (ret != 0) {
+        DPS_ERRPRINT("ERROR during handshake: %x\n", -ret);
+    }
+
+    uint32_t verifyFlags = mbedtls_ssl_get_verify_result(&cn->ssl);
+    if (verifyFlags != 0) {
+        DPS_ERRPRINT("TLSHandshake() failure when getting the verify result");
```

Comment:
TLSHandshake() now returns a value to indicate whether there was a failure, that then will be handled by deleting the remote.